Testing: Improve manually running autotests; #4047

bziemons opened the pull request:
- Testing: Move tools/run_autotest.sh to donkeyrider.py
- Testing: Cleanup additional containers on termination
- Testing: Enable filtering of test cases
- Testing: Remove --cache-from argument as it works differently nowadays.
  With docker, this needs --build-arg BUILDKIT_INLINE_CACHE=1 and with podman it only works with images built with --cache-to beforehand.

Example usage: `tools/run_autotests.sh --build -s remote_dbs --db postgres14 -d centos7 -f test_replica.py::TestReplicaCore`

I will also update the documentation on https://github.com/rucio/documentation/blob/main/docs/contributing.md

Fixes #4047 

For Your Eminence, @maany to review.

bziemons committed:
Testing: Move tools/run_autotest.sh to donkeyrider; #4047

bziemons committed:
Testing: Cleanup additional containers on termination; #4047

bziemons committed:
Testing: Enable filtering of test cases; #4047

Example usage: tools/run_autotests.sh --build -s remote_dbs --db postgres14 -d centos7 -f test_replica.py::TestReplicaCore

maany commented on etc/docker/test/matrix.yml:
```diff
@@ -16,14 +16,7 @@ dists:
-        - client_syntax
-        - remote_dbs
-        - multi_vo
+  - "3.6"
```
```
- id: 3.6
```

bziemons replied:
doesn't matter, if the entry is not a dict, it will be interpreted as the id: https://github.com/rucio/rucio/blob/2dfb00ef4e032867170aa5b02abb4272e594a9dd/tools/test/matrix_parser.py#L33-L34

I just reverted https://github.com/rucio/rucio/commit/9b5d43ca0a288bdbd3dd97573d5a082464af4081#diff-3273b199b4da4a866dddefe08c4fc04f3d69243f9472b929669ac5cc0364f608R16-R23

bziemons commented on tools/run_tests_docker.sh:
```diff
@@ -37,11 +37,11 @@ do
     i) init_only="true";;
     r) activate_rse="true";;
-    s) special="true";;
+    s) selection="true";TESTS="test_dirac.py";;
```
Needs special=true as well

joeldierkes commented on tools/test/donkeyrider.py:
```diff
@@ -0,0 +1,133 @@
+            sys.exit(1)
+
+    project_dir = os.path.abspath(os.path.join(os.path.abspath(os.path.dirname(__file__)), "../.."))
+    print("Detected project directory", project_dir, file=sys.stderr)
```
Writing to stderr as debug does not make a lot of sense from my POV. How about using `logging`?

bziemons replied:
Using stderr as logging is very common in posix environments. I don't think it is necessary to use logging for a few lines of logging inbetween. This is a developer tool after all.

joeldierkes commented on tools/test/donkeyrider.py:
```diff
@@ -0,0 +1,133 @@
+        matrix_file = os.path.join(project_dir, "etc/docker/test/matrix.yml")
+    else:
+        matrix_file = args.test_matrix_file
+    print("Using test matrix from file", matrix_file, file=sys.stderr)
```
How about setting `os.path.join(project_dir, "etc/docker/test/matrix.yml")` as default for `args.test_matrix_file`. That way we can get rid of the `if` for good and don't need to think about that.

bziemons replied:
sure, will do.

joeldierkes commented on tools/test/donkeyrider.py:
```diff
@@ -0,0 +1,133 @@
+        matrix_file = args.test_matrix_file
+    print("Using test matrix from file", matrix_file, file=sys.stderr)
+
+    # make sure the following imports work
```
Explaining comments have some flaws: They most likely get forgotten when the code is refactored and indicate, that the code is not self explaining.

How about creating functions instead of the comments and blocks under them:
```python
def main():
    check_imports()
    args = parse_args()
    matrix = parse_martrix(args)
    images = build_images(matrix)
    run_tests(matrix, images, args.filter_tests)
```

Now the code clearly communicates the steps it takes, and you don't have to look into the `parse_args` code if `parse_martrix` has a problem.

bziemons replied:
will have another look at this part of the code. I get your point.

bziemons replied:
I have boiled down the main method to something more understandable. Of course it doesn't come close to your example, since there are more variables created and used in different parts of the code.

joeldierkes commented on tools/test/donkeyrider.py:
```diff
@@ -0,0 +1,133 @@
+
+    # matrix parsing
+
+    from matrix_parser import parse_matrix
```
What is the advantage of importing the code in each place than before everything else?

bziemons replied:
it was just a routine to port each of the different tools that weren't possible to run directly in Python. I will move the imports to the top, as it is usually done.

bziemons replied:
Just realized I need the `sys.path.append(os.path.abspath(os.path.dirname(__file__)))`: That is why I don't like putting these imports to the imports on top of the file. It basically means that every other Python module just importing this module will modify the global sys.path. Running code on import is bad practice imo.

joeldierkes commented on tools/test/donkeyrider.py:
```diff
@@ -0,0 +1,133 @@
+
+def shutdown(signum, frame):
+    global stop
+    stop = True
```
The `stop` variable is used nowhere else, what is the point of it?

bziemons replied:
leftover. Will remove

joeldierkes commented on tools/test/donkeyrider.py:
```diff
@@ -0,0 +1,133 @@
+if __name__ == "__main__":
+    signal.signal(signal.SIGINT, shutdown)
+    signal.signal(signal.SIGTERM, shutdown)
+    main()
```
The name `donkeyrider` is very creative, and I like it, however it is even more un-descriptive. I would not expect someone new to the project asking themself: what is this and can I eat it?

How about `run_test_pipeline` or `run_test_suite` or ...

bziemons replied:
As a developer tool I think it is fine. I was thinking the same that it would be obscure, but somehow we sticked to it. I don't think it is necessary to change the name

bziemons committed:
Review changes requested by @joeldierkes

maany commented on tools/run_tests_docker.sh:
```diff
@@ -138,31 +138,22 @@ if test ${init_only}; then
+if test ${selection}; then
+    echo "Running the following tests:"
+    echo $TESTS | tr " " "\n"
+    tools/pytest.sh $PYTEST_SH_ARGS $TESTS
```
Quick question @bziemons: The TESTS env var should contain 1) the full path to the test or 2) path relative to the tests dir?

bziemons replied:
relative path, afaik (see the example in the PR post). It is how pytest interprets it and pytest is run on the test directory.

bziemons committed:
Testing: Remove --cache-from argument; #4047

as it works differently nowadays.
With docker, this needs --build-arg BUILDKIT_INLINE_CACHE=1 and with podman it only works with images built with --cache-to beforehand.

bziemons commented:
Rebased on master and added commit https://github.com/rucio/rucio/pull/5807/commits/91c7474705b2e115e4063db7344a7ab06ad73e40

bari12 commented:
LGTM! @bziemons please rebase it once more.
@maany any more comments?

maany reviewed (approved):
Looks good! @bziemons thanks for this :D
